Sources/App/Commands/ReconcilePackageListCommand.swift

```diff
@@ -0,0 +1,116 @@
+  
+  func run(using context: CommandContext) throws -> Future<Void>
+  {
+    // Grab the GitHub master package list and the database's current package list
+    let masterPackageList = try fetchMasterPackageList(context)
+    let currentPackageList = try fetchCurrentPackageList(context)
+    
+    // Reconcile the package lists and add/remove any new/deleted packages
+    return masterPackageList.and(currentPackageList).flatMap { (master, current) in
+      let reconciler = PackageListReconciler(masterPackageList: master, currentPackageList: current)
+      
+      return context.container.withPooledConnection(to: .psql) { database in
+        // Add any new packages
+        let packageAdditions = reconciler.packagesToAdd.map { url -> Future<Package> in
+          let package = Package(url: url)
+          return package.create(on: database)
+        }
+        
+        // Delete any removed packages
+        let packageDeletions = reconciler.packagesToDelete.map { url in
+          return Package.findByUrl(on: database, url: url).map { package -> Future<Void> in
+            print("Deleting \(package.url)")
+            return package.delete(on: database).map {
+              print("Deleted \(package.url)")
+            }
+          }
+        }
+
+        // Queue up all of the additions and deletions for execution
+        return packageAdditions.flatten(on: database)
+          .and(packageDeletions.flatten(on: database))
+          .transform(to: ())
+      }
+    }
+  }
+  
+  func fetchMasterPackageList(_ context: CommandContext) throws -> Future<[URL]>
+  {
+    return try context.container.client()
+      .get(masterPackageListURL)
+      .flatMap(to: [String].self) { response in
+        // GitHub serves the file as text/plain so it needs to be forced as JSON
+        response.http.contentType = .json
+        return try response.content.decode([String].self)
+    }.flatMap { urlStrings in
+      var urls = [URL]()
+      var errorNotifications = [Future<Void>]()
+
+      for urlString in urlStrings {
+        // This should delete/re-insert a few (5 or so) random packages per run of this command.
+        // ---
+        // This is just debugging code I put in to exercise the creation/deletion code. I was going
+        // to suggest that `ReconcilePackageListCommand` didn't really need testing since the logic
+        // gets tested in `PackageListReconciler`, but the fact that nothing is being deleted makes
+        // that look a little foolish.
+        if Int.random(in: 1...700) == 1 {
+          continue
+        }
+
+        // If this isn't a valid URL, post an error
+        guard let url = URL(string: urlString) else {
+          try errorNotifications.append(self.sendInvalidURLString(urlString, on: context))
+          continue
+        }
+
+        // Otherwise, we're all good
+        urls.append(url)
+      }
+
+      return errorNotifications
+        .flatten(on: context.container)
+        .transform(to: urls)
```

Comment:
@0xTim Thanks again for this code. Now I've had some time with it I don't understand how this line actually merges the two lists. Does `transform` merge two arrays of `Future` objects?

Reply:
`transform` takes one futures and just returns another, it's a good way to throw away futures if all you care was that they were successful. So in this case you have an array of `Future<Void>` which you wait for `flatten`. That returns `Future<[Void]>`, since we don't really care about that we can just use transform to return the array of URLs. Make sense?

Reply:
Ahhhh so it's not merging the two Future arrays, it's executing the Futures in `errorNotifications` are being executed during the `flatten`?

That makes some sense, but creates another issue in my head now :) I thought I had it figured out that *all* the Futures in a Command were executed *after* the end of the call to `run`... So, I thought that the idea was to just line up a load of Futures and then the system processes them all after `run` completes. But, if I understand your comment correctly, that's not the case and the `errorNotification` futures are being executed during `flatten`?

Reply:
Well, technically all futures in NIO are hot so they're started as soon as you start then - `flatten` just waits for them to finish